_posts/2016/2016-09-28-rest-api-test-framework.md

```diff
@@ -0,0 +1,191 @@
+                    .extract() //
+                    .body() //
+                    .jsonPath();
+{% endhighlight %}
+
+
+#### JSONPath ####
+
+REST-assured comes already with a library for JSONPath (which is basically [XPath](https://en.wikipedia.org/wiki/XPath) for JSON) namely [com.jayway.restassured:json-path](https://mvnrepository.com/artifact/com.jayway.restassured/json-path),
+that can be used to validate a JSON response and verify certain attributes.
+
+Unfortunately this library does not implement all the JSONPath expressions from [Stefan Gössner's JSONPath specification](http://goessner.net/articles/JsonPath/) - that's why we use [com.jayway.jsonpath:json-path](https://mvnrepository.com/artifact/com.jayway.jsonpath/json-path) instead.
+To use this library we cannot simply do *... .extract().body().jsonPath();* anymore. Instead we have to get the response as a string first and then parse it in a second step:
+
+{% highlight java %}
+import com.jayway.jsonpath.DocumentContext;
+import com.jayway.jsonpath.JsonPath;
+
+// ...
+
+String body = requestSpecification.get("products")
+            .then().assertThat().statusCode(200)
+            .extract().body().asString();
+
+DocumentContext products = JsonPath.parse(body);
+{% endhighlight %}
+
+Using this library we can finally do some more powerful things like:
+
+{% highlight java %}
+int itemCount = products.read("items.length()");
+// or
+List<String> productNames = products.read("items[*].name");
+{% endhighlight %}
+
+
+
+### Serenity BDD
+
+Similar to [Cucumber](https://cucumber.io/) in the Ruby world,
+Serentiy allows you to describe and structure the test cases in way that acts like a complete specification that describes each use case step by step.
+
+The following snippet shows a test case we use to check if an uploaded image was added to the slideshow of a a product:
+
+{% highlight java %}
+@RunWith(SerenityRunner.class)
+public class WhenUploadingProductImage {
+
+    private static final String IMAGE_NAME = "scrum-values-800x800.jpg";
+
+    @Steps(uniqueInstance=true)
+    ProductImageSteps steps;
+
+    @Before
+    public void setup() {
+        steps.given_a_demo_shop();
+        steps.given_a_developer_token();
+    }
+
+    @After
+    public void teardown() {
+        steps.invalidate_shop();
+    }
+
+    @Test
+    public void uploaded_image_is_added_to_slideshow() {
+        steps.given_the_first_product();
+        steps.when_uploading_a_product_image(IMAGE_NAME);
+        steps.then_the_product_slideshow_contains_image(IMAGE_NAME);
+        steps.when_the_product_image_is_changed_with_slideshow_image(IMAGE_NAME);
+        steps.then_the_product_image_is_changed_accordingly();
+    }
+
+}
+{% endhighlight %}
+
+The certain test steps are in a separate classes. We have a class *ShopSteps* which contains tests that are shop-specific and needed in all tests. Then we have e.g. *ProductSteps* inherting from ShopSteps for steps that are used in product-related tests.
```

Comment:
vielleicht besser "The implementation of the test steps can be found in different classes. " ?